pkg/commands/docker.go

```diff
@@ -66,9 +72,113 @@ func (c *DockerCommand) NewCommandObject(obj CommandObject) CommandObject {
 	return defaultObj
 }
 
+func handleSSHHosts(c *client.Client) error {
+	ctx := context.Background()
+	u, err := url.Parse(os.Getenv("DOCKER_HOST"))
+	if err != nil {
+		// if no or an invalid docker host is specified, continue nominally
+		return nil
+	}
+
+	// if the docker host scheme is "ssh", forward the docker socket before creating the client
+	if u.Scheme == "ssh" {
+		newDockerHost, err := tunneledDockerHost(ctx, u.Host)
+		if err != nil {
+			return fmt.Errorf("tunnel ssh docker host: %w", err)
+		}
+		return client.WithHost(newDockerHost)(c)
+	}
+	return nil
+}
+
+func tunneledDockerHost(ctx context.Context, remoteHost string) (string, error) {
+	socketDir, err := ioutil.TempDir("/tmp", "lazydocker-sshtunnel-")
+	if err != nil {
+		return "", fmt.Errorf("create ssh tunnel tmp file: %w", err)
+	}
+	localSocket := path.Join(socketDir, "dockerhost.sock")
+
+	err = tunnelSSH(ctx, remoteHost, localSocket)
+	if err != nil {
+		return "", fmt.Errorf("tunnel docker host over ssh: %w", err)
+	}
+
+	// set a reasonable timeout, then wait for the socket to dial successfully
+	// before attempting to create a new docker client
+	const socketTunnelTimeout = 8 * time.Second
+	ctx, cancel := context.WithTimeout(ctx, 8*time.Second)
+	defer cancel()
+
+	err = retrySocketDial(ctx, localSocket)
+	if err != nil {
+		return "", fmt.Errorf("ssh tunneled socket never became available: %w", err)
+	}
+
+	// construct the new DOCKER_HOST url with the proper scheme
+	newDockerHostURL := url.URL{Scheme: "unix", Path: localSocket}
+	return newDockerHostURL.String(), nil
+}
+
+// Attempt to dial the socket until it becomes available.
+// The retry loop will continue until the parent context is canceled.
+func retrySocketDial(ctx context.Context, socketPath string) error {
+	t := time.NewTicker(1 * time.Second)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-t.C:
+		}
+		// attempt to dial the socket, exit on success
+		err := tryDial(ctx, socketPath)
+		if err != nil {
+			continue
+		}
+		return nil
+	}
+}
+
+// Try to dial the specified unix socket, immediately close the connection if successfully created.
+func tryDial(ctx context.Context, socketPath string) error {
+	var dialer net.Dialer
+	conn, err := dialer.DialContext(ctx, "unix", socketPath)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+	return nil
+}
+
+func tunnelSSH(ctx context.Context, host, localSocket string) error {
+	cmd := exec.CommandContext(ctx, "ssh", "-L", localSocket+":/var/run/docker.sock", host, "-N")
+	// when lazydocker exits, kill the SSH tunnel
+	cmd.SysProcAttr = &syscall.SysProcAttr{
```

Comment:
I'm not sure that we actually need to do this: when the parent process is killed, any children processes should also be killed.

Reply:
After removing this, I'm seeing the tunnel process leak of macOS, seen by the following command after opening and closing `lazydocker` with `<control-c>`.
```console
ps -ax | grep sshtunnel
```

Reply:
damn. Do we know if, with the SysProcAttr line still here, the child process is killed upon quitting via 'q'? If not, I wonder if we should have a pointer to this cmd living at the top level (against the gui or the app struct) which then gets killed explicitly upon quitting the program

Reply:
I just pushed a change that improves the correctness, and has behaved properly during my manual testing of each exit method (CTR-C, q).

https://github.com/jesseduffield/lazydocker/pull/268/commits/892fc090b62bbd9f11156abc5da8201cfe540a1b